### PR TITLE
Project Auth by Max and Sandrine

### PR DESCRIPTION
**Pull request by MindstormingAB**
Link to our deployed project: https://max-sandrine-auth-app.netlify.app/
Link to our API: https://max-sandrine-auth-api.herokuapp.com/

Our reflections on how this week's project turned out: This week's project was pretty difficult because it required the new backend knowledge related to authentication we learned during the week as well as all the React/Redux skills we previously acquired during the Bootcamp. It was very challenging to build both frontend and backend from scratch.

Things I'd like to have clarified or explained in more detail: -

### CODE REVIEW

File: frontend/src/index.css [hunk 1]
```diff
@@ -5,9 +5,99 @@ body {
...
+  text-align: center;
+}
+
+a {
+  text-decoration: none;
+  padding: 1px;
+  color: white;
+  background-image: linear-gradient(#335b71 45%, #03324c 55%);
+  transition: color 0.3s, background-image 0.5s, ease-in-out;
+  /* border-radius: 25px; */
```
Comment (pwangy) [inline frontend/src/index.css:32]: Remove commented out line.

File: frontend/src/index.css [hunk 2]
```diff
@@ -5,9 +5,99 @@ body {
...
   font-family: source-code-pro, Menlo, Monaco, Consolas, "Courier New",
     monospace;
 }
+
+h1 {
+  text-align: center;
+}
+
+p {
+  margin: 0 0 16px 0;
```
Comment (pwangy) [inline frontend/src/index.css:22]: perhaps just margin-bottom?

**Review by pwangy — APPROVED** [review]
Really nice work, you two. 💪 

This week's assignment was really difficult-- and you gave us a few reminders of things we had forgotten, so thank you for that! How nice that you went the extra step to show a personalized welcome screen upon login. All those little touches you added were appreciated. 

Most of the actionable comments are minor clean-up or consistency notes; or general food-for-thought. 😁

File: backend/server.js [hunk 3]
```diff
@@ -1,29 +1,147 @@
...
+    minlength: 5
+  },
+  accessToken: {
+    type: String,
+    default: () => crypto.randomBytes(128).toString("hex"),
+    unique: true
+  }
+});
+
+userSchema.pre("save", async function (next) {
```
Comment (KarolCislowski) [inline backend/server.js:40]: nice use of  `pre()`

File: backend/server.js [hunk 4]
```diff
@@ -1,29 +1,147 @@
...
 
 // Defines the port the app will run on. Defaults to 8080, but can be 
 // overridden when starting the server. For example:
 //
 //   PORT=9000 npm start
-const port = process.env.PORT || 8080
-const app = express()
+const port = process.env.PORT || 8080;
+const app = express();
+const listEndpoints = require("express-list-endpoints");
```
[the remark this answers is not in the document]
Comment (KarolCislowski) [reply backend/server.js:68]: You should import it on top of your script with `import/from` syntax.

File: backend/server.js [hunk 5]
```diff
@@ -1,29 +1,147 @@
...
 
 // Add middlewares to enable cors and json body parsing
-app.use(cors())
-app.use(bodyParser.json())
+const allowedDomains = ["https://max-sandrine-auth-app.netlify.app", "http://localhost:3000"];
+app.use(cors({
+  origin: function (origin, callback) {
+    if (!origin) return callback(null, true);
+    if (allowedDomains.indexOf(origin) === -1) {
+      var msg = `This site ${origin} does not have an access. Only specific domains are allowed to access it`
```
Comment (KarolCislowski) [inline backend/server.js:76]: Nice that you implement this logic, but remember that you shouldn't use `var` any more. `const` would be proper here. 🤓

File: backend/server.js [hunk 6]
```diff
@@ -1,29 +1,147 @@
...
+      email: req.user.email,
+      accessToken: req.user.accessToken,
+      alias: req.user.alias
+    })
+  };
+});
+
+// Update endpoint
+app.post("/users/:userId/profile", authenticateUser);
+app.post("/users/:userId/profile", async (req, res) => {
```
Comment (KarolCislowski) [inline backend/server.js:129]: It is not a good practice to pass any user data within a URL. You should use the request body to pass it to the server. 🤓

File: frontend/src/reducers/user.js [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+      state.login.alias = alias;
+    },
+    deleteAccessToken: (state, action) => {
+      state.login.accessToken = "";
+    },
+    deleteUserId: (state, action) => {
+      state.login.userId = "";
+    },
+  }
+});
```
Comment (KarolCislowski) [inline frontend/src/reducers/user.js:34]: Reducer is a great place to manage your users but you need to store some data in the local storage to make them stay legged in after they refresh the page 🤓

**Review by KarolCislowski — CHANGES_REQUESTED** [review]
It is a great app! Good job 🦄 

You need to fix only one issue with storing your data in the local storage.

Good luck 🌈

**Comment by MindstormingAB** [comment]
Hi @KarolCislowski and @pwangy 
Thank you for reviewing this repository.
I did the following changes with regard to your feedback:
- added a local storage in frontend
- moved userId from the endpoint url to a header
- replaced POST with PATCH
- fixed smaller "bits and bobs"

**Review by KarolCislowski — APPROVED** [review]
Great work!

**Pull request closed without merging**